### PR TITLE
proxyd: Support per-RPC rate limits

### PR DESCRIPTION
**Pull request by mslipper**
Adds support for custom rate limits on a per-RPC method basis. We need this in order to limit abuse of expensive requests like `eth_getLogs`.

### CODE REVIEW

File: proxyd/server.go [hunk 1]
```diff
@@ -197,22 +213,37 @@ func (s *Server) HandleRPC(w http.ResponseWriter, r *http.Request) {
...
+		if lim == nil {
+			return false
 		}
+
+		_, _, _, ok, _ := lim.Take(ctx, xff)
+		return !ok
 	}
-	if !ok {
+
+	if isLimited("") {
```
Comment (tynes) [inline proxyd/server.go:246]: Is the empty string here correct?

Comment (tynes) [reply proxyd/server.go:246]: This is saying that if the non method override is limited correct?

File: proxyd/server.go [hunk 2]
```diff
@@ -347,6 +378,16 @@ func (s *Server) handleBatchRPC(ctx context.Context, reqs []json.RawMessage, isB
 			continue
 		}
 
+		// Take rate limit for specific methods.
+		// NOTE: eventually, this should apply to all batch requests. However,
+		// since we don't have data right now on the size of each batch, we
+		// only apply this to the methods that have an additional rate limit.
```
Comment (Inphi) [inline proxyd/server.go:384]: Should it matter if rate limiting isn't applied to batch requests that don't contain method overrides? I could be wrong, but isn't the global rate limiter meant only for downstream requests? Which we already do in `handleRPC`.

File: proxyd/server.go (hunk 1 above)
Comment (mslipper) [reply proxyd/server.go:246]: Yes - this selects the global limiter.

**Comment by Inphi** [comment]
excellent!

**Review by Inphi — APPROVED** [review]

**Review by tynes — APPROVED** [review]

**Pull request merged**